Feature: Adding thread manager

eapg opened the pull request:
* Adding Thread Manager 

This manager will be implement to create, start, kill and know the status of a thread.

reynoldmorel reviewed (changes requested):
This is better!

Now let's adapt it to our needs:

1. Rename `ThreadManager` to `AppEngineProcessor`.
2. Create an `AppEngineProcessConfig` that wiill contain any process config such as duration/timeout/interval, listeners like `on_init()`, `on_refresh()`, `before_shutdown()`, and whatever you want to add in order to setup your process. 
3. Create an `AppContext` that will contain all cross-shared information between processes
4. Create the `KitchenSimulatorProcess` and make the `AppEngineProcessor` register it.
5. Once you run `AppEngineProcessor.start()` it should start all registered processes (i.e `KitchenSimulatorProcess`).

eapg commented:
* Added EngineAppProcessor
* Added EngineAppProcessHandler to handle process to be trigger when an specific event occur
* create 3 methods inside EngineAppProcesso add_subscribers(subscriber), remove_subscriber(subscriber) and star_threads() to trigger the AppEngineprocessHandler and start all the threads stored in the process handler.

reynoldmorel commented on src/tests/utils/fixtures/app_processor_config_fixture.py:
```diff
@@ -0,0 +1,13 @@
+    id=None, interval=None, on_start=None, on_destroy=None, on_execute=None
+):
+    app_engine_config = AppProcessorConfig()
+    app_engine_config.id = id
```
Add a default value here

eapg replied:
Done

reynoldmorel commented on src/tests/core/engine/app_engine_processor_test.py:
```diff
@@ -0,0 +1,25 @@
+class TestAppEngineProcessor(unittest.TestCase):
+    @patch("src.core.engine.app_engine_processor.AppProcessorConfig")
+    @patch("src.core.engine.app_engine_processor.KitchenSimulator")
+    def test_mock_in_app_context(self, mock_kitchen, mock_config):
```
Make `mock_config` to return a `build_app_engine_config` when the instance is created. Remove the mock for KitchenSimulator.

Then assert that `len(app_engine_processor.app_context.processors) = 1` and `app_engine_processor.app_context.processors[0].app_processor_config.id` is equal to the id youi used here `build_app_engine_config`

eapg replied:
Done.

reynoldmorel commented on src/tests/core/engine/app_engine_processor_test.py:
```diff
@@ -0,0 +1,25 @@
+        "src.core.engine.app_engine_processor.KitchenSimulator.start",
+        wraps=KitchenSimulator,
+    )
+    def test_start_in_context_run_once(self, mock_kitchen):
```
This test is not ok, you should mock `KitchenSimulator` and make it to return your own instance using a `build_kitchen_simulator_processor` function and then check if the start from that mock was called when you ran `app_engine_processor.start()`

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+    def test_app_engine_config(self):
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+        self.assertEqual(test_class.app_processor_config.id, app_engine_config.id)
```
You can do this assertion in the tes above

eapg replied:
done

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        test_class = ChildProcessor(app_engine_config)
+        self.assertEqual(test_class.app_processor_config.id, app_engine_config.id)
+
+    def test_app_engine_processor_context(self):
```
Same here

eapg replied:
Done.

reynoldmorel commented on src/tests/core/engine/app_engine_processor_integration_test.py:
```diff
@@ -0,0 +1,17 @@
+        "src.core.engine.app_engine_processor.KitchenSimulator.process",
+        wraps=KitchenSimulator,
+    )
+    def test_process_run(self, mock_process):
```
Rename `mock_process`  to `mock_kitchen_simulator_process`

eapg replied:
Done. Renamed to mocked_kitchen_simulator_process

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+            test_class.app_context.processors, app_engine_processor_context.processors
+        )
+
+    def test_self_pause_is_false(self):
```
Same here

eapg replied:
Done.

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        test_class = ChildProcessor()
+        self.assertEqual(test_class.pause, False)
+
+    def test_self_destroy_is_false(self):
```
same for this one

eapg replied:
Done.

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+        test_class.start()
+        self.assertAlmostEqual(test_class.last_execution_time, datetime.now())
```
Mock `datetime.now()` and use your own fixed number.

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+
+        with patch.object(ChildProcessor, "process") as mock:
```
Could you follow same pattern as before?

eapg replied:
Done.

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        test_class.start()
+        self.assertTrue(test_class.is_alive())
+        test_class.destroy = True
+        time.sleep(15)
```
hmm this is tricky, could you mock `time.sleep` implementation and control how much time sleep function will last? It is better for scalability and flexibility

reynoldmorel commented on src/tests/core/engine/processors/abstract_processor_test.py:
```diff
@@ -0,0 +1,125 @@
+        app_engine_config = build_app_engine_config(id="test config", interval=10)
+        test_class = ChildProcessor(app_engine_config)
+
+        with patch.object(ChildProcessor, "process") as mock:
```
Use annotation as previously, or use this for every test

eapg replied:
Done.

reynoldmorel reviewed (changes requested):
Rename every `mock_*` variable to `mocked_*`.

This a nice first try to these kinf of tests.

Thanks
